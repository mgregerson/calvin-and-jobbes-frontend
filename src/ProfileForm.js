import React, { useState } from "react";
import { useContext } from "react";
import userContext from "./userContext";

/** ProfileForm
 *
 * Props:
 *       - handleProfileEdit: function passed down from App.
 *
 * State:
 *       - formData: Object with Form Data ({username, firstName, lastName, email})
 *
 * Renders a form so that a current user can edit their profile.
 */

function ProfileForm({ handleProfileEdit }) {
  // USE CONTEXT OF USER: SET CURR FORM DATA TO USERNAME, FIRSTNAME, LASTNAME, EMAIL
  const { user } = useContext(userContext);
  const [formData, setFormData] = useState({
    username: user.username,
    firstName: user.firstName,
    lastName: user.lastName,
    email: user.email,
  });

  console.log(user, "USER IN PROFILEFORM")

  /** Handles keystrokes in searchbar and updates formData */
  function handleChange(evt) {
    const fieldName = evt.target.name;
    const value = evt.target.value;

    setFormData((currData) => {
      currData[fieldName] = value;
      return { ...currData };
    });
  }
  /** handleSubmit of the form. Call function in props */
  function handleSubmit(evt) {
    evt.preventDefault();
    const updateData = formData;
    delete updateData.username;
    console.log(formData, "THE FORM DATA IN PROFILE");
    handleProfileEdit(user.username, updateData);
  }

  return (
    <div className="ProfileForm mb-4 d-flex pt-4">
      <h1 className="ProfileForm-Message">Update Profile</h1>
      <form className="ProfileForm-Form" onSubmit={handleSubmit}>
        <div className="row justify-content-center justify-content-lg-start gx-1">
          <div className="col-8">
            <label htmlFor="username">Username</label>
            <input
              id="username"
              name="username"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.username}
              aria-label="username"
              disabled
            />
            <label htmlFor="firstName">First Name</label>
            <input
              id="first-name"
              name="firstName"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.firstName}
              required
              aria-label="firstName"
            />
            <label htmlFor="lastName">Last Name</label>
            <input
              id="last-name"
              name="lastName"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.lastName}
              aria-label="lastName"
              required
            />
            <label htmlFor="email">Email</label>
            <input
              id="email"
              name="email"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.email}
              aria-label="email"
              required
            />
            <div className="col-auto">
              <button className="btn search-btn btn-lg btn-primary">
                Submit
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  );
}

export default ProfileForm;
