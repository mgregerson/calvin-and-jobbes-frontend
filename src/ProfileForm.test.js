import { render, fireEvent, screen } from "@testing-library/react";
import { useContext } from "react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import ProfileForm from "./ProfileForm";
import userContext from "./userContext";

const testHandleProfileEdit = jest.fn();

describe("ProfileForm Component", () => {
  const mockUser = {
    username: "test-username",
    firstName: "test-firstName",
    lastName: "test-lastName",
    email: "test-email",
  };

  const WrapperComponent = ({ children }) => (
    <userContext.Provider value={{ user: mockUser }}>
      {children}
    </userContext.Provider>
  );

  it("renders without crashing", () => {
    render(
      <MemoryRouter>
        <WrapperComponent>
          <ProfileForm handleProfileEdit={testHandleProfileEdit} />
        </WrapperComponent>
      </MemoryRouter>
    );
  });

  it("matches snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <WrapperComponent>
          <ProfileForm handleProfileEdit={testHandleProfileEdit} />
        </WrapperComponent>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("Dynamically updates the value of the search box with current input", function () {
    const { container } = render(
      <MemoryRouter>
        <WrapperComponent>
          <ProfileForm handleProfileEdit={testHandleProfileEdit} />
        </WrapperComponent>
      </MemoryRouter>
    );

    const inputIds = ["first-name", "last-name", "email"];
    const inputFields = inputIds.map((id) => container.querySelector(`#${id}`));

    for (let field of inputFields) {
      fireEvent.change(field, { target: { value: "a" } });
      expect(field.value).toEqual("a");
      fireEvent.change(field, { target: { value: "an" } });
      expect(field.value).toEqual("an");
      fireEvent.change(field, { target: { value: "" } });
      expect(field.value).toEqual("");
    }
  });

  it("Allows a user to update their profile with valid inputs", function () {
    const { container } = render(
      <MemoryRouter>
        <WrapperComponent>
          <ProfileForm handleProfileEdit={testHandleProfileEdit} />
        </WrapperComponent>
      </MemoryRouter>
    );

    const inputIds = ["first-name", "last-name", "email"];
    const inputFields = inputIds.map((id) => container.querySelector(`#${id}`));

    for (let field of inputFields) {
      fireEvent.change(field, { target: { value: "newtest@test.com" } });
      expect(field.value).toEqual("newtest@test.com");
    }
    const submitBtn = container.querySelector(".search-btn");
    fireEvent.click(submitBtn);
    expect(testHandleProfileEdit).toHaveBeenCalled();
  });

  it("Does not allow a user to update their username", function () {
    const { container } = render(
      <MemoryRouter>
        <WrapperComponent>
          <ProfileForm handleProfileEdit={testHandleProfileEdit} />
        </WrapperComponent>
      </MemoryRouter>
    );
    const usernameInput = container.querySelector(`#username`);
    expect(usernameInput).toBeDisabled();
    expect(usernameInput).toHaveValue("test-username");
  });

  it("Does not allow a user to update their email to a non-email type", function () {
    const { container } = render(
      <MemoryRouter>
        <WrapperComponent>
          <ProfileForm handleProfileEdit={testHandleProfileEdit} />
        </WrapperComponent>
      </MemoryRouter>
    );
    const emailInput = container.querySelector("#email");

    fireEvent.change(emailInput, { target: { value: "notavalidemail" } });
    expect(emailInput.value).toEqual("notavalidemail");

    const submitBtn = container.querySelector(".search-btn");
    fireEvent.click(submitBtn);

    // The form will not process, so the user is not navigated to a new page.
    expect(screen.queryByText("Update Profile")).toBeInTheDocument();
  });
});
