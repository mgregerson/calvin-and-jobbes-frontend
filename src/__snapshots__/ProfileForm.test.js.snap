// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm Component matches snapshot 1`] = `
<div>
  <div
    class="pt-5"
  >
    <div
      class="ProfileForm col-md-6 col-lg-4 offset-md-3 offset-lg-4"
    >
      <h3
        class="ProfileForm-message"
      >
        Update Profile
      </h3>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form
            class="ProfileForm-Form"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              aria-label="username"
              class="form-control form-control-lg"
              disabled=""
              id="username"
              name="username"
              value="test-username"
            />
            <label
              for="firstName"
            >
              First Name
            </label>
            <input
              aria-label="firstName"
              class="form-control form-control-lg"
              id="first-name"
              name="firstName"
              placeholder="First Name"
              required=""
              value="test-firstName"
            />
            <label
              for="lastName"
            >
              Last Name
            </label>
            <input
              aria-label="lastName"
              class="form-control form-control-lg"
              id="last-name"
              name="lastName"
              placeholder="Last Name"
              required=""
              value="test-lastName"
            />
            <label
              for="email"
            >
              Email
            </label>
            <input
              aria-label="email"
              class="form-control form-control-lg"
              id="email"
              name="email"
              placeholder="E-Mail"
              required=""
              type="email"
              value="test-email"
            />
            <div
              class="ProfileForm-button"
            >
              <button
                class="btn search-btn btn-lg btn-primary"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
